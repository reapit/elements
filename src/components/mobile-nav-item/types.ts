import { ReactNode } from 'react'

type CommonMobileNavItemProps = {
  label: string
  isActive?: boolean
  hasBadge?: boolean
}

export interface MobileNavItemSimpleProps extends CommonMobileNavItemProps {
  children?: never
  isDefaultExpanded?: never
  href: string
}

export interface MobileNavItemExpandableProps extends CommonMobileNavItemProps {
  children: ReactNode
  isDefaultExpanded?: boolean
  href?: never
}

export type MobileNavSubItemProps = MobileNavItemSimpleProps
