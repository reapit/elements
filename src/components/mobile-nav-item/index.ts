export * from './mobile-nav-item'
export * from './styles'
