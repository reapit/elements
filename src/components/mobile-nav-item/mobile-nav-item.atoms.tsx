import { useState } from 'react'
import type { FC, ReactNode, MouseEventHandler } from 'react'

import { useId } from '../../storybook/random-id'

import { Icon } from '../icon'

import {
  ElMobileNavItemAnchor,
  ElMobileNavItemBadge,
  ElMobileNavItemButton,
  ElMobileNavItemContent,
  ElMobileNavItemExpanderButton,
  ElMobileNavItemListItem,
  ElMobileNavSubItemUnorderedList,
} from './styles'

type CommonMobileNavItemProps = {
  label: string
  isActive?: boolean
  hasBadge?: boolean
}

interface MobileNavItemSimpleAsButtonElementProps extends CommonMobileNavItemProps {
  onClick: MouseEventHandler<HTMLButtonElement>
  href?: never
  children?: never
}

interface MobileNavItemSimpleAsAnchorElementProps extends CommonMobileNavItemProps {
  href: string
  onClick?: never
  children?: never
}

export type MobileNavItemSimpleProps = MobileNavItemSimpleAsButtonElementProps | MobileNavItemSimpleAsAnchorElementProps

const isMobileNavItemSimpleAsButtonElement = (
  props: MobileNavItemSimpleProps,
): props is MobileNavItemSimpleAsButtonElementProps => {
  return 'onClick' in props
}

export const MobileNavItemSimple: FC<MobileNavItemSimpleProps> = (props) => {
  if (isMobileNavItemSimpleAsButtonElement(props)) {
    const { onClick, isActive, label, hasBadge, ...rest } = props ?? {}

    return (
      <ElMobileNavItemListItem {...rest} aria-label={label}>
        <ElMobileNavItemButton onClick={onClick} aria-current={isActive ? 'true' : undefined}>
          <ElMobileNavItemContent>
            {label}
            {hasBadge && <ElMobileNavItemBadge />}
          </ElMobileNavItemContent>
        </ElMobileNavItemButton>
      </ElMobileNavItemListItem>
    )
  } else {
    const { href, isActive, label, hasBadge, ...rest } = props ?? {}

    return (
      <ElMobileNavItemListItem {...rest} aria-label={label}>
        <ElMobileNavItemAnchor href={href} aria-current={isActive ? 'page' : undefined}>
          <ElMobileNavItemContent>
            {label}
            {hasBadge && <ElMobileNavItemBadge />}
          </ElMobileNavItemContent>
        </ElMobileNavItemAnchor>
      </ElMobileNavItemListItem>
    )
  }
}

export interface MobileNavItemExpandableProps extends CommonMobileNavItemProps {
  children: ReactNode
  href?: never
  onClick?: never
}

export const MobileNavItemExpandable: FC<MobileNavItemExpandableProps> = (props) => {
  const [isExpanded, setIsExpanded] = useState(Boolean(props?.isActive))
  const panelId = useId()

  const { isActive, label, hasBadge, children, ...rest } = props ?? {}

  return (
    <ElMobileNavItemListItem {...rest} data-is-expanded={isExpanded} aria-label={label}>
      <ElMobileNavItemExpanderButton
        type="button"
        aria-expanded={isExpanded ?? isActive}
        aria-current={isActive ? 'true' : undefined}
        aria-controls={panelId}
        onClick={() => setIsExpanded((prev) => !prev)}
      >
        <ElMobileNavItemContent>
          {label}
          {hasBadge && <ElMobileNavItemBadge />}
        </ElMobileNavItemContent>

        <Icon icon={isExpanded ? 'chevronUp' : 'chevronDown'} fontSize="16px" />
      </ElMobileNavItemExpanderButton>

      <ElMobileNavSubItemUnorderedList id={panelId} aria-hidden={!isExpanded}>
        {children}
      </ElMobileNavSubItemUnorderedList>
    </ElMobileNavItemListItem>
  )
}
