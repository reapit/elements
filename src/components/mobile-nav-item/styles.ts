import { styled } from '@linaria/react'

const baseStyles = `
  cursor: pointer;
  color: inherit;
  display: flex;
  border: none;
  outline: none;
  width: 100%;
  border-radius: inherit;
  border: 4px solid transparent;
  background: inherit;
  padding: var(--space-2, 8px) var(--space-4, 16px);
  font-family: var(--font-family, Inter);
  font-size: var(--font-size-base, 15px);
  font-style: normal;
  font-weight: var(--font-weight-regular, Regular);
  line-height: var(--line-height-base, 24px);
  letter-spacing: var(--letter-spacing-base, -0.15px);

  &:hover {
    background: var(--fill-default-light, #e5e9ed);
  }

  &:focus-visible {
    border: 4px solid var(--purple-300, #7e9bfa);
  }

  &:active,
  &[aria-current='true'],
  &[aria-current='page'] {
    color: var(--text-action, #4e56ea);
  }
`

export const ElMobileNavItemAnchor = styled.a`
  ${baseStyles}
`

export const ElMobileNavItemButton = styled.button`
  ${baseStyles}
`

export const ElMobileNavItemExpanderButton = styled.button`
  ${baseStyles}
`

export const ElMobileNavItemContent = styled.span`
  display: flex;
  align-items: center;
  justify-content: flex-start;
  gap: var(--space-2, 8px);
  flex-grow: 1;
`

export const ElMobileNavItemBadge = styled.span`
  display: block;
  width: var(--size-2, 8px);
  height: var(--size-2, 8px);
  background-color: var(--icon-error, #f01830);
  border-radius: 100%;
`

export const ElMobileNavSubItemUnorderedList = styled.ul`
  &[aria-hidden='true'] {
    display: none;
  }
`

export const ElMobileNavItemListItem = styled.li`
  display: flex;
  flex-direction: column;
  background: var(--fill-white, #fff);
  border-radius: var(--corner-lg, 8px);

  &[data-is-expanded='true'] {
    background: var(--fill-default-lightest, #f2f4f6);

    > ${ElMobileNavItemExpanderButton} {
      border-radius: var(--corner-lg, 8px) var(--corner-lg, 8px) var(--corner-none, 0) var(--corner-none, 0);
    }
    ${ElMobileNavSubItemUnorderedList} > * {
      background: var(--fill-default-lightest, #f2f4f6);
      border-radius: var(--corner-none, 0px);
    }
    ${ElMobileNavSubItemUnorderedList} > :last-child {
      border-radius: var(--corner-none, 0) var(--corner-none, 0) var(--corner-lg, 8px) var(--corner-lg, 8px);
    }
    ${ElMobileNavItemAnchor}, ${ElMobileNavItemExpanderButton} {
      &:hover {
        background: var(--fill-default-light, #e5e9ed);
      }
    }
  }
`
