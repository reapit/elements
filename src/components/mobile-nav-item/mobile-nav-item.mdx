import { Meta, Canvas, Controls } from '@storybook/blocks'

import { RenderHtmlMarkup } from '../../storybook/render-html-markup'

import * as MobileNavItemStories from './mobile-nav-item.stories'

<Meta of={MobileNavItemStories} />

# Mobile Nav Item

**A Versatile Navigation Component**

The `MobileNavItem` is a flexible component designed to render navigation items within a `MobileNav` context. It offers two primary variants:

**1. Simple Variant**

- Renders a basic navigation item using an `HTMLAnchorElement`

**2. Expandable Variant**

- Creates an expandable navigation item with a parent element (`HTMLButtonElement`) and a sub-menu (`HTMLUListElement`)
- The sub-menu contains `MobileNavSubItem` components, each rendered using an `HTMLAnchorElement`

## Simple Variant

**Default Appearance**

A basic `MobileNavItem` without any special styling or interactions

<Canvas of={MobileNavItemStories.Simple} />

**With Badge**

<Canvas of={MobileNavItemStories.SimpleWithBadge} />

**With Active State**

<Canvas of={MobileNavItemStories.SimpleWithActiveState} />

## Expandable Variant

**Basic Usage**

<Canvas of={MobileNavItemStories.Expandable} />

<RenderHtmlMarkup of={MobileNavItemStories.Expandable} />

**Default Expanded**

An expandable `MobileNavItem` that is initially open

<Canvas of={MobileNavItemStories.DefaultExpanded} />

<RenderHtmlMarkup of={MobileNavItemStories.DefaultExpanded} />
