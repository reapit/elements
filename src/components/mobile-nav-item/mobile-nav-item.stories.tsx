import type { Meta, StoryObj } from '@storybook/react'

import { MobileNavItem } from './mobile-nav-item'
import { action } from '@storybook/addon-actions'

const meta = {
  title: 'Components/Mobile Nav Item',
  component: MobileNavItem,
  parameters: {
    layout: 'padded',
  },
  args: {
    label: 'Item 1',
    children: null,
  },
} satisfies Meta<typeof MobileNavItem>

export default meta
type Story = StoryObj<typeof meta>

export const Simple: Story = {
  render: ({}) => <MobileNavItem label="Item 1" href="#item-1" />,
}

export const Active: Story = {
  render: ({}) => <MobileNavItem label="Item 1" href="#item-1" isActive />,
}

export const WithBadge: Story = {
  render: ({}) => <MobileNavItem label="Item 1" href="#item-1" hasBadge />,
}

export const Expandable: Story = {
  render: ({}) => (
    <MobileNavItem label="Item 2" hasBadge>
      <MobileNavItem label="Sub Item 2.1" href="#item-2.1" />
      <MobileNavItem label="Sub Item 2.2" onClick={action('onClick')} hasBadge />
      <MobileNavItem label="Sub Item 2.3" href="#item-2.3" />
    </MobileNavItem>
  ),
}

export const DefaultExpanded: Story = {
  render: ({}) => (
    <MobileNavItem label="Item 3" isActive>
      <MobileNavItem label="Sub Item 3.1" href="#item-3.1" />
      <MobileNavItem label="Sub Item 3.2" onClick={action('onClick')} isActive />
      <MobileNavItem label="Sub Item 3.3" href="#item-3.3" />
    </MobileNavItem>
  ),
}
