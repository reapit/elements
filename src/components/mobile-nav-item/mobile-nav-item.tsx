import type { FC } from 'react'

import { MobileNavItemExpandable, MobileNavItemSimple } from './mobile-nav-item.atoms'
import type { MobileNavItemExpandableProps, MobileNavItemSimpleProps } from './mobile-nav-item.atoms'

const isExpandableVariant = (props: MobileNavItemProps): props is MobileNavItemExpandableProps => {
  return 'children' in props
}

export type MobileNavItemProps = MobileNavItemSimpleProps | MobileNavItemExpandableProps

export const MobileNavItem: FC<MobileNavItemProps> = (props) => {
  if (isExpandableVariant(props)) {
    return <MobileNavItemExpandable {...props} />
  } else {
    return <MobileNavItemSimple {...props} />
  }
}
