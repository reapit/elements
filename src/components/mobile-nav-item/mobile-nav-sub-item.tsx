import type { FC } from 'react'

import { MobileNavSubItemProps } from './types'
import { ElMobileNavItemAnchor, ElMobileNavItemBadge, ElMobileNavItemContent, ElMobileNavItemListItem } from './styles'

export const MobileNavSubItem: FC<MobileNavSubItemProps> = ({ href, isActive, label, hasBadge }) => {
  return (
    <ElMobileNavItemListItem role="none">
      <ElMobileNavItemAnchor href={href} aria-current={isActive ? 'page' : undefined} role="menuitem">
        <ElMobileNavItemContent>
          {label}
          {hasBadge && <ElMobileNavItemBadge />}
        </ElMobileNavItemContent>
      </ElMobileNavItemAnchor>
    </ElMobileNavItemListItem>
  )
}
