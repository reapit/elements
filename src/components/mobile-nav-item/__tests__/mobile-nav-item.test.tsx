import { render } from '@testing-library/react'

import { MobileNavItem } from '../mobile-nav-item'

describe('MobileNavItem', () => {
  it('should render Simple MobileNavItem', () => {
    expect(render(<MobileNavItem label="Item 1" href="#item-1" />).asFragment()).toMatchSnapshot()
  })

  it('should render Simple MobileNavItem with badge', () => {
    expect(render(<MobileNavItem label="Item 1" href="#item-1" hasBadge />).asFragment()).toMatchSnapshot()
  })

  it('should render Expandable MobileNavItem', () => {
    expect(
      render(
        <MobileNavItem label="Item 2" hasBadge isActive>
          <MobileNavItem label="Sub Item 2.1" href="#item-2.1" isActive />
          <MobileNavItem label="Sub Item 2.2" onClick={() => {}} hasBadge />
          <MobileNavItem label="Sub Item 2.3" href="#item-2.3" />
        </MobileNavItem>,
      ).asFragment(),
    ).toMatchSnapshot()
  })
})
