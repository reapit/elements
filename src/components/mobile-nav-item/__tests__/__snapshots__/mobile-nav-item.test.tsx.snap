// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MobileNavItem > should render Expandable MobileNavItem 1`] = `
<DocumentFragment>
  <li
    aria-label="Item 2"
    class="mocked-styled-7 el-mobile-nav-item-list-item"
    data-is-expanded="true"
  >
    <button
      aria-controls="test-static-id"
      aria-current="true"
      aria-expanded="true"
      class="mocked-styled-3 el-mobile-nav-item-expander-button"
      type="button"
    >
      <span
        class="mocked-styled-4 el-mobile-nav-item-content"
      >
        Item 2
        <span
          class="mocked-styled-5 el-mobile-nav-item-badge"
        />
      </span>
      <span
        class="mocked-styled-0 el-icon"
        style="font-size: 16px;"
      >
        <svg
          fill="none"
          height="1em"
          role="img"
          style="font-size: 16px;"
          title="Icon image with name chevronUp"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18.9948 16.4959C18.7292 16.4974 18.4739 16.3932 18.2853 16.2062L11.9991 9.91315L5.71303 16.2062C5.32115 16.5979 4.68579 16.5979 4.29391 16.2062C3.90203 15.8145 3.90203 15.1795 4.29391 14.7878L11.2896 7.79547C11.4772 7.60637 11.7327 7.5 11.9991 7.5C12.2656 7.5 12.5211 7.60637 12.7087 7.79547L19.7044 14.7878C19.8936 14.9754 20 15.2307 20 15.497C20 15.7634 19.8936 16.0187 19.7044 16.2062C19.5158 16.3932 19.2605 16.4974 18.9948 16.4959Z"
            fill="currentColor"
          />
        </svg>
      </span>
    </button>
    <ul
      aria-hidden="false"
      class="mocked-styled-6 el-mobile-nav-sub-item-unordered-list"
      id="test-static-id"
    >
      <li
        aria-label="Sub Item 2.1"
        class="mocked-styled-7 el-mobile-nav-item-list-item"
      >
        <a
          aria-current="page"
          class="mocked-styled-1 el-mobile-nav-item-anchor"
          href="#item-2.1"
        >
          <span
            class="mocked-styled-4 el-mobile-nav-item-content"
          >
            Sub Item 2.1
          </span>
        </a>
      </li>
      <li
        aria-label="Sub Item 2.2"
        class="mocked-styled-7 el-mobile-nav-item-list-item"
      >
        <button
          class="mocked-styled-2 el-mobile-nav-item-button"
        >
          <span
            class="mocked-styled-4 el-mobile-nav-item-content"
          >
            Sub Item 2.2
            <span
              class="mocked-styled-5 el-mobile-nav-item-badge"
            />
          </span>
        </button>
      </li>
      <li
        aria-label="Sub Item 2.3"
        class="mocked-styled-7 el-mobile-nav-item-list-item"
      >
        <a
          class="mocked-styled-1 el-mobile-nav-item-anchor"
          href="#item-2.3"
        >
          <span
            class="mocked-styled-4 el-mobile-nav-item-content"
          >
            Sub Item 2.3
          </span>
        </a>
      </li>
    </ul>
  </li>
</DocumentFragment>
`;

exports[`MobileNavItem > should render Simple MobileNavItem 1`] = `
<DocumentFragment>
  <li
    aria-label="Item 1"
    class="mocked-styled-7 el-mobile-nav-item-list-item"
  >
    <a
      class="mocked-styled-1 el-mobile-nav-item-anchor"
      href="#item-1"
    >
      <span
        class="mocked-styled-4 el-mobile-nav-item-content"
      >
        Item 1
      </span>
    </a>
  </li>
</DocumentFragment>
`;

exports[`MobileNavItem > should render Simple MobileNavItem with badge 1`] = `
<DocumentFragment>
  <li
    aria-label="Item 1"
    class="mocked-styled-7 el-mobile-nav-item-list-item"
  >
    <a
      class="mocked-styled-1 el-mobile-nav-item-anchor"
      href="#item-1"
    >
      <span
        class="mocked-styled-4 el-mobile-nav-item-content"
      >
        Item 1
        <span
          class="mocked-styled-5 el-mobile-nav-item-badge"
        />
      </span>
    </a>
  </li>
</DocumentFragment>
`;
